mesonbuild/build.py

```diff
@@ -1408,7 +1408,7 @@ def link(self, target):
                 msg = f"Can't link non-PIC static library {t.name!r} into shared library {self.name!r}. "
                 msg += "Use the 'pic' option to static_library to build with PIC."
                 raise InvalidArguments(msg)
-            if self.for_machine is not t.for_machine:
+            if self.for_machine is not t.for_machine and (not t.uses_rust() or t.rust_crate_type != 'proc-macro'):
```

Comment:
I think `t` can be a `CustomTarget` which does not implement `uses_rust()`. What about:

```suggestion
            if self.for_machine is not t.for_machine and (not isinstance(t, SharedLibrary) or t.rust_crate_type != 'proc-macro'):
```